nat-lab/tests/test_pq.py

```diff
@@ -372,3 +378,106 @@ async def test_dns_with_pq(
 
         # Expect this to work as well after the secret key change
         await query_dns(client_conn, "google.com")
+
+
+@pytest.mark.parametrize(
+    "alpha_setup_params",
+    [
+        pytest.param(
+            SetupParameters(
+                connection_tag=ConnectionTag.DOCKER_CONE_CLIENT_1,
+                adapter_type_override=TelioAdapterType.NEP_TUN,
+                connection_tracker_config=generate_connection_tracker_config(
+                    ConnectionTag.DOCKER_CONE_CLIENT_1,
+                    stun_limits=ConnectionLimits(1, 1),
+                    nlx_1_limits=ConnectionLimits(1, 2),
+                ),
+                is_meshnet=False,
+            ),
+        ),
+    ],
+)
+async def test_pq_vpn_silent_pq_upgrader(
+    alpha_setup_params: SetupParameters,
+) -> None:
+    async with AsyncExitStack() as exit_stack:
+        alpha_setup_params.features.post_quantum_vpn.handshake_retry_interval_s = 1
+
+        env = await exit_stack.enter_async_context(
+            setup_environment(exit_stack, [alpha_setup_params])
+        )
+
+        client_conn, *_ = [conn.connection for conn in env.connections]
+        client, *_ = env.clients
+
+        wg_server = config.WG_SERVER  # use non PQ server
+        await client.connect_to_vpn(
+            str(wg_server["ipv4"]),
+            int(wg_server["port"]),
+            str(wg_server["public_key"]),
+            pq=True,
+        )
+
+        await client.wait_for_log("Failed to fetch PQ keys: TIMEOUT")
```

Comment:
question: what happens in the current libtelio with this testcase? there was `recv` timeout, wonder if it would just swallow the timeout or produce the error message as well?

Reply:
Semantically it should behave the same. However, there's no such log in the current released version. I've added these tests mostly to compensate for removing some tests in the `proto.rs` module

Reply:
You mean non-PQ WG connection to VPN servers? 

In that case retry logic and timeouts are handled by WireGuard implementations (neptun on apple/android, wg-nt on windows, linux-kernel-module on Linux) which generally employs constant backoff strategy (e.g. retries handshake every 5s)